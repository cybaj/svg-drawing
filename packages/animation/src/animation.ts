import {
  createSvgElement,
  createSvgChildElement,
  pathObjectToElement,
  Path,
  Svg,
  camel2kebab,
  roundUp,
} from '@svg-drawing/core'
import { AnimationOption, FrameAnimation } from './types'

export class Animation {
  public ms: number
  private _anim: FrameAnimation | null
  private _origin: Svg
  private _frames: number | undefined
  private _repeatCount: string
  constructor(public svg: Svg, { ms }: AnimationOption = { ms: 60 }) {
    this.ms = ms
    this._anim = null
    this._origin = this.generateOrigin()
    this._repeatCount = 'indefinite'
  }

  /**
   * `frame` is the number of frames to animate `repeat` is related for
   * repeatCount of animate element attribute.
   *
   * @param {FramaAnimation} fn
   * @param {{ frame?: number; repeat?: number }} opts
   */
  public setAnimation(
    fn: FrameAnimation,
    {
      frames,
      repeatCount,
      ms,
    }: { frames?: number; repeatCount?: number | string; ms?: number } = {}
  ): void {
    this._anim = fn
    this._frames = frames
    this._origin = this.generateOrigin()
    this._repeatCount = repeatCount ? `${repeatCount}` : 'indefinite'
    if (ms) this.ms = ms
  }

  public restore(): void {
    this.svg.copy(this._origin)
  }

  public generateFrame(index?: number): Path[] {
    if (!this._anim) return this.svg.paths
    return this._anim(
      this._origin.paths.map((p) => p.clone()),
      index
    )
  }

  public update() {
    this._origin = this.generateOrigin()
  }

  public toElement(): SVGSVGElement {
    const animPathsList: Path[][] = Array(this.frames)
      .fill(null)
      .map((_: any, i: number) => this.generateFrame(i))

    const dur = this.frames * (this.ms > 0 ? this.ms : 1) + 'ms'
    const t = 1 / this.frames
    const keyTimes = `0;${Array(this.frames)
      .fill(undefined)
      .map((_, i) => roundUp((i + 1) * t, 4) + '')
      .join(';')}`

    const createAnimationElement = (
      origin: Path,
      attributeName: string,
      defaultValue: string,
      getValue: ({
        origin,
        path,
      }: {
        origin: Path
        path?: Path
      }) => string | undefined
    ): SVGElement | null => {
      const animValues = animPathsList.map((ap) => {
        const path = ap.find((p) => p.attrs.id === origin.attrs.id)
        return getValue({ origin, path }) || defaultValue
      })
      // return null if value is same all.
      if (animValues.every((v) => v === defaultValue)) return null

      return createSvgChildElement('animate', {
        dur,
        keyTimes,
        attributeName,
        repeatCount: this._repeatCount,
        values: [defaultValue, ...animValues].join(';'),
      })
    }

    const animEls = this._origin.paths.map((p) => {
      const pEl = pathObjectToElement(p.toJson())
      const dAnimEl = createAnimationElement(
        p,
        'd',
        p.getCommandString(),
        ({ origin, path }) =>
          path && path.commands.length > 0
            ? path.getCommandString()
            : origin.commands[0].toString()
      )
      if (dAnimEl) pEl.appendChild(dAnimEl)

      // TODO: Check attribute key and value.
      const { id, ...attrs } = p.attrs // exclude id
      Object.entries(attrs).map(
        ([propertyName, val]: [string, string | undefined]) => {
          if (!val) return
          const attrName = camel2kebab(propertyName)
          const aEl = createAnimationElement(
            p,
            attrName,
            val,
            ({ path }) => path?.attrs[propertyName]
          )
          if (aEl) pEl.appendChild(aEl)
        }
      )

      return pEl
    })

    const size = {
      width: String(this.svg.width),
      height: String(this.svg.height),
    }
    const bgEl = this.svg.background
      ? [
          createSvgChildElement('rect', {
            ...size,
            fill: this.svg.background,
          }),
        ]
      : []
    return createSvgElement(
      {
        width: String(this.svg.width),
        height: String(this.svg.height),
      },
      bgEl.concat(animEls)
    )
  }

  /** @returns {number} Default value is total of commands length. */
  private get frames(): number {
    return this._frames && this._frames > 0
      ? this._frames
      : this._origin.paths.reduce((l, p) => l + p.commands.length, 0)
  }

  private generateOrigin(): Svg {
    const svg = this.svg.clone()
    svg.paths.map((p, i) => p.updateAttributes({ id: `t${i}` }))
    return svg
  }
}
