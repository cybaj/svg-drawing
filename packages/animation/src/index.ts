export * from './SvgAnimation'
export * from './animation'
export * from './types'
