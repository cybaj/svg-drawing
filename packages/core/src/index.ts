export * from './drawing'
export * from './convert'
export * from './renderer'
export * from './svg'
export * from './throttle'
export * from './utils'
export * from './handler'
export * from './types'
export * from './download'
