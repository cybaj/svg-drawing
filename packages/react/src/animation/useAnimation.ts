import { Animation, AnimateObject } from '@svg-drawing/animation'
import { useCallback, useState, useMemo } from 'react'
import type { Path } from '@svg-drawing/core'

type UseAnimation = (arg: { paths: Path[] }) => [
  AnimateObject | null,
  {
    update: () => void
    clear: () => void
    setup: Animation['setup']
  },
  Animation
]

export const useAnimation: UseAnimation = ({ paths }) => {
  const [animateObject, setAnimateObject] = useState<AnimateObject | null>(null)

  const animation = useMemo(() => new Animation(), [])

  const setup = useCallback<Animation['setup']>(
    (frame, opts) => {
      animation.setup(frame, opts)
    },
    [animation]
  )

  const update = useCallback(() => {
    animation.initialize(paths)
    setAnimateObject(animation.toJson())
  }, [animation, paths])

  const clear = useCallback(() => {
    setAnimateObject(null)
  }, [])

  return [
    animateObject,
    {
      setup,
      update,
      clear,
    },
    animation,
  ]
}
