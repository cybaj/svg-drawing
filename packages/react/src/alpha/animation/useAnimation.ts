import { useCallback, useRef } from 'react'
import { Animation } from '@svg-drawing/animation'
import { Svg } from '@svg-drawing/core/lib'

export const useAnimation = ({
  ms,
  sharedSvg,
  background,
}: {
  ms: number
  sharedSvg: Svg
  background: string
}): Animation['setAnimation'] => {
  const targetRef = useRef<HTMLDivElement | null>(null)

  const setupAnimation = useCallback(
    (...arg: Parameters<Animation['setAnimation']>) => {
      if (!targetRef.current) return

      const animation = new Animation(sharedSvg)
      animation.setAnimation(...arg)

      const animationSvg = animation.toElement()
      targetRef.current.replaceChild(
        animationSvg,
        targetRef.current.childNodes[0]
      )
    },
    [sharedSvg]
  )

  return setupAnimation
}
