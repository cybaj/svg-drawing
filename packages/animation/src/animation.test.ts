import { parseSVGString } from '@svg-drawing/core'
import { Animation } from './animation'

const defaultTestData = `<svg width="200" height="200">
  <path fill="#f00" stroke="#00f" stroke-width="4" d="M 1 1 L 2 2 C 3 3 5 3 7 3 Z"></path>
  <path fill="#ff0" stroke="#f0f" stroke-width="2" d="M 2 2 L 4 4 C 6 6 10 6 14 6 Z"></path>
</svg>`

describe('Animation', () => {
  const init = (svgStr = defaultTestData) =>
    new Animation().initialize(parseSVGString(svgStr))
  it('new Animation()', () => {
    expect(init()).toMatchSnapshot()
  })
  it('generateFrame', () => {
    const anim = init()
    anim.setAnimation((paths) => [paths[0]])

    expect(anim.generateFrame()).toMatchObject([anim.origin.paths[0]])
  })
  // TODO: Improve test pattern
  it('toElement', () => {
    const anim = init()
    anim.setAnimation((paths, count) => {
      const update = []
      for (let i = 0; i < paths.length; i += 1) {
        // Test property
        if (count % 2 === 0) paths[i].updateAttributes({ stroke: '#0ff' })
        // Test Attribute
        if (count % 3 === 0)
          paths[i].updateAttributes({
            strokeLinecap: 'mitter',
          })
        if (count < paths[i].commands.length) {
          paths[i].commands = paths[i].commands.slice(0, count)
          update.push(paths[i])
          break
        }
        count -= paths[i].commands.length
        update.push(paths[i])
      }
      return update
    })

    expect(anim.toElement()).toMatchSnapshot()
  })
})
