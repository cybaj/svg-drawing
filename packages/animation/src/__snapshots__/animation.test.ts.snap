// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animation new Animation() 1`] = `
Animation {
  "_anim": null,
  "_repeatCount": "indefinite",
  "ms": 60,
  "origin": Svg {
    "background": undefined,
    "height": 200,
    "paths": Array [
      Path {
        "attrs": Object {
          "fill": "#f00",
          "id": "t0",
          "stroke": "#00f",
          "strokeLinecap": "round",
          "strokeWidth": "4",
        },
        "commands": Array [
          Move {
            "points": Array [
              Point {
                "_x": 1,
                "_y": 1,
              },
            ],
            "type": "M",
          },
          Line {
            "points": Array [
              Point {
                "_x": 2,
                "_y": 2,
              },
            ],
            "type": "L",
          },
          Curve {
            "points": Array [
              Point {
                "_x": 3,
                "_y": 3,
              },
              Point {
                "_x": 5,
                "_y": 3,
              },
              Point {
                "_x": 7,
                "_y": 3,
              },
            ],
            "type": "C",
          },
          Close {
            "type": "z",
          },
        ],
      },
      Path {
        "attrs": Object {
          "fill": "#ff0",
          "id": "t1",
          "stroke": "#f0f",
          "strokeLinecap": "butt",
          "strokeWidth": "2",
        },
        "commands": Array [
          Move {
            "points": Array [
              Point {
                "_x": 2,
                "_y": 2,
              },
            ],
            "type": "M",
          },
          Line {
            "points": Array [
              Point {
                "_x": 4,
                "_y": 4,
              },
            ],
            "type": "L",
          },
          Curve {
            "points": Array [
              Point {
                "_x": 6,
                "_y": 6,
              },
              Point {
                "_x": 10,
                "_y": 6,
              },
              Point {
                "_x": 14,
                "_y": 6,
              },
            ],
            "type": "C",
          },
          Close {
            "type": "z",
          },
        ],
      },
    ],
    "width": 200,
  },
}
`;

exports[`Animation toElement 1`] = `
<svg
  height="200"
  version="1.1"
  width="200"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M1 1 L2 2 C3 3 5 3 7 3 z"
    fill="#f00"
    id="t0"
    stroke="#00f"
    stroke-linecap="round"
    stroke-width="4"
  >
    <animate
      attributeName="d"
      dur="480ms"
      keyTimes="0;0.125;0.25;0.375;0.5;0.625;0.75;0.875;1"
      repeatCount="indefinite"
      values="M1 1;M1 1;M1 1 L2 2;M1 1 L2 2 C3 3 5 3 7 3;M1 1 L2 2 C3 3 5 3 7 3 z;M1 1 L2 2 C3 3 5 3 7 3 z;M1 1 L2 2 C3 3 5 3 7 3 z;M1 1 L2 2 C3 3 5 3 7 3 z;M1 1 L2 2 C3 3 5 3 7 3 z"
    />
    <animate
      attributeName="stroke"
      dur="480ms"
      keyTimes="0;0.125;0.25;0.375;0.5;0.625;0.75;0.875;1"
      repeatCount="indefinite"
      values="#0ff;#00f;#0ff;#00f;#0ff;#00f;#0ff;#00f;#00f"
    />
  </path>
  <path
    d="M2 2 L4 4 C6 6 10 6 14 6 z"
    fill="#ff0"
    id="t1"
    stroke="#f0f"
    stroke-linecap="butt"
    stroke-width="2"
  >
    <animate
      attributeName="d"
      dur="480ms"
      keyTimes="0;0.125;0.25;0.375;0.5;0.625;0.75;0.875;1"
      repeatCount="indefinite"
      values="M2 2;M2 2;M2 2;M2 2;M2 2;M2 2;M2 2 L4 4;M2 2 L4 4 C6 6 10 6 14 6;M2 2 L4 4 C6 6 10 6 14 6 z"
    />
    <animate
      attributeName="stroke"
      dur="480ms"
      keyTimes="0;0.125;0.25;0.375;0.5;0.625;0.75;0.875;1"
      repeatCount="indefinite"
      values="#f0f;#f0f;#f0f;#f0f;#0ff;#f0f;#0ff;#f0f;#f0f"
    />
  </path>
</svg>
`;
